dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/constant/DefaultConstantService.java

```diff
@@ -106,7 +140,8 @@ public Constant getConstant( String uid )
     @Transactional( readOnly = true )
     public List<Constant> getAllConstants()
     {
-        return constantStore.getAll();
```

Comment:
@jimgrace I don't like using caches for the general get methods. The problem here is that the Hibernate session will be lost, and _if_ constant had some associations to other entities, we would run into LazyLoadExceptions. We normally don't do this in the general service layer, so the caller has no way of knowing he is getting a cached, detached collection. It is also breaking from the standard convention.

The cache invalidation will also not work, as we are using the metadata importer to create constants, not the service layer directly.

I would instead just create the cache of constants where the cached instances are needed.

Reply:
Thanks, @larshelge I've done that now. Ironically that's exactly what I did at first for Ghana. The only downside of course is if constants are changed the cache is stale, which is why I tried putting the code in the constant service and using the general caching service. But as you note that won't work and can lead to other problems.

I've now reverted the initial commits, and instead supplied the same code that fixed Ghana's problem in a custom dhis.war that we supplied to them. This code caches the constantMap as needed in the DefaultExpressionService.

Reply:
Great, thanks @jimgrace , looks good.

Related, @mortenoh found that the transaction readOnly=true parameter on ConstantService might be the root cause here. We have backported a fix there now to 2.34 and 2.35. Not sure what version Ghana is on.

I agree constants could be cached in general in ExpressionService, they should not change often (as the name implies). I think it is fine to use the DHIS 2 Cache interface here as you had it initially, and just cache for an hour or so. That will be a bit more standard, so that we don't have yet another way of caching things in the system. What do you think?

Reply:
@larshelge thanks. I've refactored from Supplier to DHIS2 Cache. Ghana is on 2.35.1 + my patches, the rest of which have now been backported to 2.35. Let's see if we can get them on 2.35.2. :)

I agree that for end users, changing constant values 1 hour after add/update should be fine. I was thinking more of the metadata developers who add/update a constant, whether on a production system or a test system. The next thing they'll likely do is try the constant in an expression. If it's inexplicably wrong for up to 2 minutes this may be less likely to cause them serious trouble than for an hour, especially if they don't understand what's going on. Two minutes solves Ghana's performance problem. It's too bad that we have to design for the end user or the metadata developer but not both. I could still change it, but I wonder whether the performance improvement would be worth the metadata developer confusion (and our support of them).

I liked the Java 8 Supplier interface because it did exactly what I wanted: cache only one object, define the getting function upfront, then just call supplier.get() each time for the value. If it could be useful elsewhere in DHIS2, we could write a wrapper around our Cache to provide a Supplier interface. We could also hide the awkwardness of the dummy key ("x") in the wrapper so the end user class would not have to use it.